src/KaratsubaMatrices.jl

```diff
@@ -2,32 +2,224 @@ struct KMat{MatrixType,Int}
+            bits_match = match(r"\d+", string(type))
+            bits = parse(Int, bits_match.match) - 1
+        elseif occursin("Int", string(type))
+            bits_match = match(r"\d+", string(type))
+            bits = parse(Int, bits_match.match)
+        else
+            error("The input type is neither Int, UInt, nor Float.")
+        end
+    
+        if bits == -1
+            error("Input type is not recognized.")
+        end
+        if N >= BigInt(2)^bits
+            error("Modulus too large")
+        end
+
+        return new{Matrix{eltype(A)},Int}(A,B,N)
+    end
 end
 
 struct KMatMulPlan{MatrixType}
     temp1::MatrixType
     temp2::MatrixType
 end
 
-function KMatMul!(C,A,B,plan)
-    plan.temp1 = A.l + A.h
-    plan.temp2 = B.l + B.h
-    C.l = A.l*B.l
-    C.h = plan.temp1*plan.temp2
-    C.h = C.h - C.l
-    C.h = C.h - A.h*B.h
+function find_max_ops(type, N)
+
+    if occursin("Float", string(type))
+        bits_dict = Dict("64" => 51, "32" => 22, "16" => 9)
+        bits_match = match(r"\d+", string(type))
+        bits = get(bits_dict, bits_match.match, -1)
+    elseif occursin("UInt", string(type))
+        bits_match = match(r"\d+", string(type))
+        bits = parse(Int, bits_match.match) - 1
+    elseif occursin("Int", string(type))
+        bits_match = match(r"\d+", string(type))
+        bits = parse(Int, bits_match.match)
+    else
+        error("The input type is neither Int, UInt, nor Float.")
+    end
+
+    if bits == -1
+        error("Input type is not recognized.")
+    end
+
+    if 64 ≤ bits
+        floor(BigInt, (BigInt(2)^bits - 1) / N^2) - 1
+    else
+        floor(Int, (2^bits - 1) / N^2) - 1    
+    end
+
+    
+end
+
+function KMatMul!(C::KMat,A::KMat,B::KMat,plan::KMatMulPlan)
+    if (A.m != B.m) || (A.m != C.m)
+        error("Matrices must have the same modulus m")
+    end
+    if ncols(A.l) != nrows(B.l)
+        error("Matrix dimensions don't work for multiplication")
+    end
+    if (nrows(A.l) != nrows(C.l)) || (ncols(B.l) != ncols(C.l))
+        error("Output matrix has wrong dimensions")
+    end
+
+    max_ops = find_max_ops(eltype(A.l),max(size(A.l)...,size(B.l)...))
+
+    if A.m > max_ops
+        error("Modulus is larger than max_ops")
+    end
+
+    plan.temp1 .= A.l + A.h
+    plan.temp2 .= B.l + B.h
+    C.l .= A.l*B.l
+    C.h .= plan.temp1*plan.temp2
+    C.h .= C.h - C.l
+    C.h .= C.h - A.h*B.h
+end
+
+function KMatToMat(K::KMat)
+    A = zeros(eltype(K.l),nrows(K.l),ncols(K.l))
+    A = K.l + K.m*K.h
+    return A
 end
 
-function KMatToMat(K)
-    A = zeros(eltyple(K.l),nrows(K.l),ncols(K.l))
+function KMatToMat(T,K)
+    A = zeros(T,nrows(K.l),ncols(K.l))
     A = K.l + K.m*K.h
     return A
 end
 
 function MatToKMat(A,M)
     K = KMat{Matrix{eltype(A)},Int}(zeros(eltype(A),nrows(A),ncols(A)),zeros(eltype(A),nrows(A),ncols(A)),M)
-    K.h = trunc.(A./M)
-    K.l = A - K.h
-    K.m = M
+    K.h .= trunc.(A./M)
+    K.l .= A - K.m*K.h
+    return K
```

Comment:
For what it's worth, Julia automatically returns the value of the last line of any function. So typing `return` isn't necessary.